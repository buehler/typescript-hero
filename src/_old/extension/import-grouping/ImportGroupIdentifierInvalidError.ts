/**
 * Thrown when a import group identifier neither does match a keyword nor a regex pattern.
 * 
 * @export
 * @class ImportGroupIdentifierInvalidError
 * @extends {Error}
 */
export class ImportGroupIdentifierInvalidError extends Error {
    constructor(identifier: string) {
        super();
        this.message = `The identifier "${identifier}" does not match a keyword or a regex pattern (/ .. /).`;
    }
}
