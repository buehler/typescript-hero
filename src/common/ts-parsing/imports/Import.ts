import { Generatable } from '../ts-generation';
import { Clonable } from '../utilities';
import { Node } from './Node';

/**
 * Basic import interface. Defines an import in a document.
 * If no start and end points are given, the import is considered "new".
 * 
 * @export
 * @interface Import
 */
export interface Import extends Clonable, Node, Generatable {
    /**
     * Indicates if the given import is "new" or a parsed one.
     * 
     * @type {boolean}
     * @memberOf Import
     */
    readonly isNew: boolean;

    /**
     * The library name of the given import (This actually is the part after the import statement).
     *
     * @example "express"
     *
     * @type {string}
     * @memberOf Import
     */
    libraryName: string;
}

/**
 * Basic interface for aliased imports. Defines an alias for namespace imports and other aliased imports.
 * 
 * @export
 * @interface AliasedImport
 * @extends {Import}
 */
export interface AliasedImport extends Import {
    /**
     * Alias for the given import. E.g. for a "* as ..." import.
     * 
     * @type {string}
     * @memberOf AliasedImport
     */
    alias: string;
}
