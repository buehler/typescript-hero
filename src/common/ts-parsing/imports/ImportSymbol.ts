import { Generatable } from '../../ts-generation';
import { Clonable } from '../../utilities';

/**
 * Class for symbols that are contained in a named import.
 * (i.e. import {SYMBOL} from '...').
 * 
 * @export
 * @class ImportSymbol
 * @implements {Clonable}
 */
export class ImportSymbol implements Clonable, Generatable {
    constructor(public specifier: string, public alias?: string) { }

    /**
     * Generates typescript code out of the actual import.
     * 
     * @returns {string}
     * 
     * @memberOf DefaultImport
     */
    public generateTypescript(): string {
        return `${this.specifier}${this.alias ? ` as ${this.alias}` : ''}`;
    }

    /**
     * Clones the current resolve specifier and returns a new instance with the same properties.
     * 
     * @returns {ImportSymbol}
     * 
     * @memberOf ImportSymbol
     */
    public clone(): ImportSymbol {
        return new ImportSymbol(this.specifier, this.alias);
    }
}
