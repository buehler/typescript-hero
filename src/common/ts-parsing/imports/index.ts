export * from './DefaultImport';
export * from './ExternalModuleImport';
export * from './Import';
export * from './NamedImport';
export * from './NamespaceImport';
export * from './StringImport';
