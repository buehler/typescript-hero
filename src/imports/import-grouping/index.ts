export * from './import-group';
export * from './import-group-identifier-invalid-error';
export * from './import-group-keyword';
export * from './import-group-order';
export * from './import-group-setting-parser';
export * from './keyword-import-group';
export * from './regex-import-group';
export * from './remain-import-group';
