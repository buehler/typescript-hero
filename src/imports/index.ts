export { ImportAppender } from './import-appender';
export { ImportOrganizer } from './import-organizer';
