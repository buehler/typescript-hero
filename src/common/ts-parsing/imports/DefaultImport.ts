import { GenerationOptions } from '../../ts-generation';
import { AliasedImport, importRange } from './Import';
import { Serializable } from 'ts-json-serializer';
import { Range, TextDocument } from 'vscode-languageserver-types';

/**
 * Default import. Imports the default exports of a file.
 * (i.e. import foobar from ...).
 * 
 * @export
 * @class DefaultImport
 * @implements {AliasedImport}
 */
@Serializable({ factory: json => new DefaultImport(json.libraryName, json.alias, json.start, json.end) })
export class DefaultImport implements AliasedImport {
    public get isNew(): boolean {
        return this.start !== undefined && this.end !== undefined;
    }

    constructor(public libraryName: string, public alias: string, public start?: number, public end?: number) { }

    /**
     * Generates typescript code out of the actual import.
     * 
     * @param {GenerationOptions} {stringQuoteStyle, eol}
     * @returns {string}
     * 
     * @memberOf DefaultImport
     */
    public generateTypescript({stringQuoteStyle, eol}: GenerationOptions): string {
        return `import ${this.alias} from ${stringQuoteStyle}${this.libraryName}${stringQuoteStyle}${eol}\n`;
    }

    /**
     * Calculates the document range of the node in the given document.
     * 
     * @param {TextDocument} document
     * @returns {Range}
     * 
     * @memberOf DefaultImport
     */
    public getRange(document: TextDocument): Range {
        return importRange(document, this.start, this.end);
    }

    /**
     * Clone the current import object.
     * 
     * @returns {DefaultImport}
     * 
     * @memberOf DefaultImport
     */
    public clone(): DefaultImport {
        return new DefaultImport(this.libraryName, this.alias, this.start, this.end);
    }
}
