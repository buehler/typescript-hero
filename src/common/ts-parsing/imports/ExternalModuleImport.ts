import { GenerationOptions } from '../../ts-generation';
import { DocumentLike } from '../Node';
import { AliasedImport, importRange } from './Import';
import { Serializable } from 'ts-json-serializer';
import { Range } from 'vscode-languageserver-types';

/**
 * Alternative to the namespace import. Can be used by various libraries.
 * (i.e. import foobar = require('foobar')).
 * 
 * @export
 * @class ExternalModuleImport
 * @implements {AliasedImport}
 */
@Serializable({ factory: json => new ExternalModuleImport(json.libraryName, json.alias, json.start, json.end) })
export class ExternalModuleImport implements AliasedImport {
    public get isNew(): boolean {
        return this.start !== undefined && this.end !== undefined;
    }

    constructor(public libraryName: string, public alias: string, public start?: number, public end?: number) { }

    /**
     * Generates typescript code out of the actual import.
     * 
     * @param {GenerationOptions} {stringQuoteStyle, eol}
     * @returns {string}
     * 
     * @memberOf ExternalModuleImport
     */
    public generateTypescript({stringQuoteStyle, eol}: GenerationOptions): string {
        return `import ${this.alias} = require(${stringQuoteStyle}${this.libraryName}${stringQuoteStyle})${eol}\n`;
    }

    /**
     * Calculates the document range of the node in the given document.
     * 
     * @param {DocumentLike} document
     * @returns {Range}
     * 
     * @memberOf ExternalModuleImport
     */
    public getRange(document: DocumentLike): Range {
        return importRange(document, this.start, this.end);
    }

    /**
     * Clone the current import object.
     * 
     * @returns {ExternalModuleImport}
     * 
     * @memberOf ExternalModuleImport
     */
    public clone(): ExternalModuleImport {
        return new ExternalModuleImport(this.libraryName, this.alias, this.start, this.end);
    }
}
