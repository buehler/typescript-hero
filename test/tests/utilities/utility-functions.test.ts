import { join } from 'path';
import { TypescriptParser } from 'typescript-parser';
import { workspace } from 'vscode';

import { ImportGroupKeyword, KeywordImportGroup, RegexImportGroup } from '../../../src/imports/import-grouping';
import ioc from '../../../src/ioc';
import iocSymbols from '../../../src/ioc-symbols';
import {
  getImportInsertPosition,
  importGroupSortForPrecedence,
  importSortByFirstSpecifier,
} from '../../../src/utilities/utility-functions';
import { expect } from '../setup';

describe('utility functions', () => {

  describe('getImportInsertPosition()', () => {

    class MockDocument {
      constructor(private documentText: string) { }

      public getText(): string {
        return this.documentText;
      }
    }

    it('should return top position if no editor is specified', () => {
      const pos = getImportInsertPosition(undefined);
      expect(pos).to.matchSnapshot();
    });

    it('should return the top position if empty file', () => {
      const pos = getImportInsertPosition({
        document: new MockDocument(''),
      } as any);
      expect(pos).to.matchSnapshot();
    });

    it('should return correct position for commented file', () => {
      const pos = getImportInsertPosition({
        document: new MockDocument('    // This is a file header\nStart of file\n'),
      } as any);
      expect(pos).to.matchSnapshot();
    });

    it('should return correct position for use strict', () => {
      const pos = getImportInsertPosition({
        document: new MockDocument(`'use strict'\nStart of file\n`),
      } as any);
      expect(pos).to.matchSnapshot();
    });

    it('should return correct position for jsdoc comment open', () => {
      const pos = getImportInsertPosition({
        document: new MockDocument('/** start of a jsdoc\n'),
      } as any);
      expect(pos).to.matchSnapshot();
    });

    it('should return correct position for jsdoc comment line', () => {
      const pos = getImportInsertPosition({
        document: new MockDocument(' * jsdoc line\n'),
      } as any);
      expect(pos).to.matchSnapshot();
    });

    it('should return correct position for jsdoc comment close', () => {
      const pos = getImportInsertPosition({
        document: new MockDocument('*/\n'),
      } as any);
      expect(pos).to.matchSnapshot();
    });

  });

  describe('importGroupSortForPrecedence', () => {

    it('should prioritize regexes, leaving original order untouched besides that', () => {
      const initialList = [
        new KeywordImportGroup(ImportGroupKeyword.Modules),
        new KeywordImportGroup(ImportGroupKeyword.Plains),
        new RegexImportGroup('/cool-library/'),
        new RegexImportGroup('/cooler-library/'),
        new KeywordImportGroup(ImportGroupKeyword.Workspace),
      ];
      const expectedList = initialList
        .slice(2, 4)
        .concat(initialList.slice(0, 2))
        .concat(initialList.slice(4));

      expect(importGroupSortForPrecedence(initialList)).to.deep.equal(
        expectedList,
        'Regex Import Groups should appear first (and that should be the only change)',
      );
    });

  });

  describe('importSortByFirstSpecifier', () => {
    const parser = ioc.get<TypescriptParser>(iocSymbols.parser);
    const rootPath = workspace.workspaceFolders![0].uri.fsPath;

    it('should sort according to first specifier/alias, falling back to module path', async () => {
      const file = await parser.parseFile(
        join(
          rootPath,
          'utilities',
          'imports-for-specifier-sort.ts',
        ),
        rootPath,
      );

      const result = [...file.imports].sort(importSortByFirstSpecifier);
      expect(result.map(i => i.libraryName)).to.matchSnapshot();
    });
  });

});
