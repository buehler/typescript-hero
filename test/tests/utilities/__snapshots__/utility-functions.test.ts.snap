// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utility functions getImportInsertPosition() should return correct position for commented file 1`] = `
Object {
  "character": 0,
  "line": 1,
}
`;

exports[`utility functions getImportInsertPosition() should return correct position for jsdoc comment close 1`] = `
Object {
  "character": 0,
  "line": 1,
}
`;

exports[`utility functions getImportInsertPosition() should return correct position for jsdoc comment line 1`] = `
Object {
  "character": 0,
  "line": 1,
}
`;

exports[`utility functions getImportInsertPosition() should return correct position for jsdoc comment open 1`] = `
Object {
  "character": 0,
  "line": 1,
}
`;

exports[`utility functions getImportInsertPosition() should return correct position for use strict 1`] = `
Object {
  "character": 0,
  "line": 1,
}
`;

exports[`utility functions getImportInsertPosition() should return the top position if empty file 1`] = `
Object {
  "character": 0,
  "line": 0,
}
`;

exports[`utility functions getImportInsertPosition() should return top position if no editor is specified 1`] = `
Object {
  "character": 0,
  "line": 0,
}
`;

exports[`utility functions importSortByFirstSpecifier should sort according to first specifier/alias, falling back to module path 1`] = `
Array [
  "./anotherFile",
  "coolEffectLib",
  "./myFile",
  "myLib",
  "anotherLib",
  "./workspaceSideEffectLib",
]
`;
